### ISSUE TITLE
tf.io.gfile.listdir does not with with S3 buckets

### ISSUE THREAD

**Issue by orf** [tensorflow/tensorflow#36893]
**System information** - MacOS, Tensorflow 2.1, Python 3.7.6

**Describe the current behavior**

Executing `tf.io.gfile.listdir("s3://bucket-name/")` results in:

```
/private/var/folders/bn/36k9myl51_v23yyxm629ly4w0000gn/T/tmp.PygT1otw/lib/python3.7/site-packages/tensorflow_core/python/lib/io/file_io.py in list_directory_v2(path)
    644   return [
    645       compat.as_str_any(filename)
--> 646       for filename in pywrap_tensorflow.GetChildren(compat.as_bytes(path))
    647   ]
    648

InvalidArgumentError: S3 path doesn't contain an object name: s3://bucket-name/
```

**Describe the expected behavior**

It should return a list of objects within the bucket.

S3 is quite limited in features for searching/globbing S3 objects but simple operations like listing objects under a given key (in the case above, `/`), are well supported and it would be good to include in Tensorflow.

The error message is also somewhat confusing.

**Comment by ravikyram** [comment]
@orf 
Could you please share simple stand alone code to reproduce the issue in our environment. It helps in localizing the issue faster. Thanks!

**Comment by orf** [comment]
Hey @ravikyram , so I provided it in the description: `tf.io.gfile.listdir("s3://bucket-name/")`.

You obviously need a S3 bucket set up. I can provide this privately if required, but I expect you have the ability to create one yourselves.

**Comment by 0x0badc0de** [comment]
The same stuff. Have to use `s3://bucket/data` instead of just `s3://bucket/` to store files and for globing to work.

**Comment by jean-lucas** [comment]
Any updates on this?

**Comment by Saduf2019** [comment]
@orf 
Is this still an issue.

**Comment by mihaimaruseac** [comment]
I think this will get fixed in the 2.4 final release, after [rewriting implementation of matching code](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/file_system_helper.cc;l=123-266;drc=8b5b9dc96666a3a5d27fad7179ff215e3b74b67c)
